tactics/Goalie_2.py

```diff
@@ -0,0 +1,510 @@
+
+		if state:
+			state=state.stateB	
+		# self.gtB = GoToBall.GoToBall(BOT_BALL_THRESH)
+		# self.gtB.add_kub(self.kub)
+		# self.gtB.add_theta(theta=normalize_angle(atan2(target.y - state.ballPos.y,target.x - state.ballPos.x)))
+		# self.gtB.spin()
+
+		#Will Use KickToPoint after making it perfect with kicking
+
+
+		self.ktp = KickToPoint.KickToPoint(target)
+		self.ktp.add_kub(self.kub)
+		self.ktp.add_theta(theta=normalize_angle(atan2(target.y - state.ballPos.y,target.x - state.ballPos.x)))
+		self.ktp.spin()
+
+
+	def execute_active(self):
+		print("entered the active state")
+		state = None
+		rospy.wait_for_service('bsServer',)
+		getState = rospy.ServiceProxy('bsServer',bsServer)
+		try:
+			state = getState(state)
+		except rospy.ServiceException, e:
+			print("exception found")
+
+		if state:
+			state=state.stateB
+
+
+		if opponent_with_the_ball(state)!=None:
+			if state.ballVel.x < -10:
+				angle = math.atan(state.awayPos[state.opp_bot_closest_to_ball].theta)
+				expected_y = angle*(-HALF_FIELD_MAXX + 3*BOT_RADIUS - state.ballPos.x) + state.ballPos.y
+			else:
+				if abs(state.ballPos.y) < OUR_GOAL_MAXY:
+					expected_y = state.ballPos.y
+				else:
+					if state.ballPos.y < 0:
+						expected_y = OUR_GOAL_MINY
+					else:
+						expected_y = OUR_GOAL_MAXY
+			#print("Entered Protect , going to", point.x, point.y)
+			point = Vector2D(-HALF_FIELD_MAXX + 5*BOT_RADIUS,expected_y)
+		
+			#expected_y = angle*(-HALF_FIELD_MAXX - state.ballPos.x) + state.ballPos.y
+			start_time = rospy.Time.now()
+			start_time = 1.0*start_time.secs + 1.0*start_time.nsecs/pow(10,9)   
+			generatingfunction = _GoToPoint_.execute(start_time,BOT_BALL_THRESH)
+			for gf in generatingfunction:
+				self.kub,aim = gf
+				print("gf", point.x, point.y)
+				if not vicinity_points(aim,point,thresh=BOT_RADIUS) or self.outOfField(state):
+					self.behavior_failed = True
+					print("Fail Fail")
+					break
+
+		elif state.ballPos.x<-HALF_FIELD_MAXX+3*BOT_RADIUS:
+			target = Vector2D(0,0)
+			state = None
+			rospy.wait_for_service('bsServer',)
+			getState = rospy.ServiceProxy('bsServer',bsServer)
+			try:
+				state = getState(state)
+			except rospy.ServiceException, e:
+				print("exception found")
+
+			if state:
+				state=state.stateB	
+
+			self.ktp = KickToPoint.KickToPoint(target)
+			self.ktp.add_kub(self.kub)
+			self.ktp.add_theta(theta=normalize_angle(atan2(target.y - state.ballPos.y,target.x - state.ballPos.x)))
+			self.ktp.spin()
+
+	def execute_protect(self):
+		# pass
+		state = None
+		rospy.wait_for_service('bsServer',)
+		getState = rospy.ServiceProxy('bsServer',bsServer)
+		try:
+			state = getState(state)
+		except rospy.ServiceException, e:
+			print("exception found")
+
+		if state:
+			state=state.stateB	
+		#expected_y = goalie_expected_y(state, self.kub.kubs_id)
+		if state.ballVel.x < -10:
+			#angle = state.ballVel.y/state.ballVel.x  
+			angle = math.atan(state.awayPos[state.opp_bot_closest_to_ball].theta)
```

Comment:
Also it is better not to use state.opp_bot_closest_to_ball here as it was removed as an unwanted belief state recently by @shamin1998 .  @shamin1998  do we still have opp_bot_closest_to_ball variable in BS? Is the computation reliable?

Reply:
I'd removed the variables that weren't being used at the time. But I guess we'll need opp_bot_closest_to_ball for role selector, I'll put it back in